src/app/epics/get-items.epic.ts

```diff
@@ -23,19 +24,21 @@ export class GetItemsEpic {
         private searchHistoryActions: SearchHistoryActions,
     ) {}
 
-    getDefaultItems = (action$): Observable<AnyAction> => {
+    getDefaultItems = (action$: ActionsObservable<AnyAction>): Observable<AnyAction> => {
         return action$.ofType(ItemsActions.GET_DEFAULT_ITEMS)
             .do( (action: AnyAction) => console.log('Epic sees action: ', action.type) )
             .mergeMap( () => this.appService.getTrendingItems() )
             .map((items:IItemsRef) => this.itemsActions.recievedDefaultItems(items, Date.now()) );
     }
 
-    getItemsByQuery = (action$): Observable<AnyAction> => {
+    getItemsByQuery = (action$: ActionsObservable<AnyAction>): Observable<AnyAction> => {
         const queryAndItems$: Observable<[string, IItemsRef]> = action$
             .ofType(ItemsActions.GET_ITEMS_BY_QUERY)
             .map( (action: AnyAction) => action.query )
             .switchMap( (query: string) => 
-                this.appService.getItemsByQuery(query).map( (items: IItemsRef) => [query, items] ) 
+                this.appService.getItemsByQuery(query).map( (items: IItemsRef) => 
+                    [query, items] as [string, IItemsRef] 
+                ) 
```

Comment:
This epic is a bit funny

Depending on your intent, there could be different ways to handle it.

First question being, is it important that the search query is stored only _after_ the items query request resolves?
If not, then I'd remove that concern from this epic and simply handle that at the reducer level since the reducer will also have `ItemsActions.GET_ITEMS_BY_QUERY` run through it.

If it is important, I like that you've discovered that you can emit multiple actions from an epic - it can be a useful tool.
We may be able to handle this in a bit more of a concise/declarative manner however.

```
.switchMap( (query: string) => 
  this.appService.getItemsByQuery(query).map( (items: IItemsRef) => [query, items] ) 
```

Rather than stringing along `query` and making it the concern of the `getItemsByQuery` to map it back out as an array, we can leverage the second argument to `switchMap` called a project function (other operators have this same argument). It allows us to "map" or "project" a new value resulting from the current emission, and whatever the operators emission would be.

We could rewrite the switchMap as...
```
.switchMap(
  (query: string) => this.appService.getItemsByQuery(query),
  (query, items) => ({(query, items, date: Date.now()}))
)
```

What we're doing now is collecting the query and items emissions and mapping them into an object (an array would work fine as well)

Since you also want date to be available across the two resulting actions, let's include that here now in a single place rather than duplicating code.

Now both resulting actions can leverage the exact same time

Additionally, rather than creating two extra streams and then returning the `Observable.merge`, we can instead switchMap to an new observable that simply emits two actions.

Something like...

```
.switchMap(
  (query: string) => this.appService.getItemsByQuery(query),
  (query, items) => ({(query, items, date: Date.now()}))
)
.switchMap(({ query, items, date }) => Observable.from([
  this.itemsActions.recievedItemsByQuery(items, date)),
  this.searchHistoryActions.addSearchTerm(query, date)
]))
```

Reply:
I do want the search history to be stored AFTER items are received. I couldn't think of how to write what you have more elegantly -- it's nice. 

This was the first Epic I wrote, so retrospectively I would actually feel better if the `addSearchTerm` action was Dispatched from a separate Epic that watches for the `receivedItemsByQuery` action. I would refactor my Redux State so that upon `getItemsByQuery` action dispatch, the query is saved as the `currentQuery` in the redux state. That way, I avoid the search history reducer from handling the get items by query actions case, all together. Does that sound like a better approach? (I don't think I'll refactor it as such at this point...maybe later on)